velox/functions/prestosql/aggregates/PrestoHasher.cpp

```diff
@@ -231,45 +231,63 @@ void PrestoHasher::hash<TypeKind::ROW>(
     BufferPtr& hashes) {
   auto baseRow = vector_->base()->as<RowVector>();
   auto indices = vector_->indices();
-  SelectivityVector elementRows;
 
-  if (vector_->isIdentityMapping()) {
+  SelectivityVector elementRows;
+  if (vector_->isIdentityMapping() && !vector_->mayHaveNulls()) {
     elementRows = rows;
   } else {
     elementRows = SelectivityVector(baseRow->size(), false);
-    rows.applyToSelected(
-        [&](auto row) { elementRows.setValid(indices[row], true); });
+    rows.applyToSelected([&](auto row) {
+      if (!vector_->isNullAt(row)) {
+        elementRows.setValid(indices[row], true);
+      }
+    });
     elementRows.updateBounds();
   }
 
   BufferPtr childHashes =
       AlignedBuffer::allocate<int64_t>(elementRows.end(), baseRow->pool());
 
   auto rawHashes = hashes->asMutable<int64_t>();
-  auto rowChildHashes = childHashes->as<int64_t>();
 
   if (isTimestampWithTimeZoneType(vector_->base()->type())) {
     // Hash only timestamp value.
     children_[0]->hash(baseRow->childAt(0), elementRows, childHashes);
+    auto rawChildHashes = childHashes->as<int64_t>();
     rows.applyToSelected([&](auto row) {
       if (!baseRow->isNullAt(indices[row])) {
-        rawHashes[row] = rowChildHashes[indices[row]];
+        rawHashes[row] = rawChildHashes[indices[row]];
       } else {
         rawHashes[row] = 0;
       }
     });
     return;
   }
 
+  BufferPtr combinedChildHashes =
+      AlignedBuffer::allocate<int64_t>(elementRows.end(), baseRow->pool());
+  auto* rawCombinedChildHashes = combinedChildHashes->asMutable<int64_t>();
+  std::fill_n(rawCombinedChildHashes, rows.end(), 1);
+
   std::fill_n(rawHashes, rows.end(), 1);
 
   for (int i = 0; i < baseRow->childrenSize(); i++) {
     children_[i]->hash(baseRow->childAt(i), elementRows, childHashes);
```

Comment:
Should we reset childHashes , before computing for every child. 
Can we have a child vector that has the null bit set at a point but the parent doesnt have the null bit set, in that case we will not compute hash at that point for child and reuse and older hash value which might be wrong. 

Let me know if this is not possible.

Reply:
The expectation is that hash() function will compute hashes for all rows requested regardless whether they are null or not.